### ISSUE TITLE
add image to drumkit properties

### ISSUE THREAD

**Issue by thijz** [hydrogen-music/hydrogen#212]
adding a picture to the drumkit properties makes it much more appealing for the user to use
something like this would be nice : 

![mockup_image_in_drumkit_library](https://cloud.githubusercontent.com/assets/1384210/5347703/77ec8efa-7f2a-11e4-8537-517c4b7e2764.jpg)
![mockup_image_in_drumkit_properties](https://cloud.githubusercontent.com/assets/1384210/5347704/77ecbe3e-7f2a-11e4-84a5-89f754f3e762.jpg)

**Comment by pvint** [comment]
Is this being worked on by anyone?

I've been wanting something do to contribute, and this sounds like a reasonably easy way to get my feet wet!

**Comment by thijz** [comment]
hi pvint
there is nobody working on this one, so go ahead and get your feet wet :-)

before you get started it's probably best to check out how drumkuts are packaged 
basically these are tgz files that contain all samples + an XML that describes the drumkit.  so i guess the picture would also have to go in that tgz

some time ago i created a py script that creates drumkits, so this might give you some insight on the way drumkits are structured (see http://audio-and-linux.blogspot.com/2012/10/hydrogen-drumkit-creator-script-v5-is.html)

important note : it's best to clone the hydrogen git repo and create small patches for what you are working on.  if anything would break it is easier to track down where things went wrong (maybe of this particular feature it wont be necessary to split up, but for other, bigger, features it it the way to go)

all info on getting the code and building it can be found on the wiki : 
https://github.com/hydrogen-music/hydrogen/wiki

good luck and dont hesitate to ask questions (best via the mailing list)

grtz
Thijs

**Comment by pvint** [comment]
I have this partially done, I added an <image> tag to the xml with the filename of the image, and it displays in the properties dialog much like your image above, and is copied to a new drumkit correctly when the kit is saved under a new name.

It is *not* in the Instruments->Save dialog yet, nor in the library browser, but I thought I'd let you have a look before I go too much further. 

You can have a look at the diffs at https://github.com/hydrogen-music/hydrogen/compare/master...pvint:master or I can do a pull request if you'd like.

Also note: The TR-808 drumkit is updated and it get's rid of the XML errors.

Question: Is the source for feeds/drumkit_list.php available somewhere?

**Comment by thijz** [comment]
Hi Paul
Yesterday i built hydrogen from your fork but i didnt have time to play
around with it yet
I'll have a go during my lunch break

Grtz
Thijs
Op 25-mrt.-2015 16:52 schreef "Paul Vint" <notifications@github.com>:

**Comment by thijz** [comment]
Hi Paul

just gave it a quick try and looks nice :-)
i guess some min/max size/resolution/dimensions constraints will have to be
defined to guarantee decent quality
is it possible to use a png with transparency  ?  (or an other format that
supports transparency)

grtz
Thijs

2015-03-27 7:56 GMT+01:00 Thijs van severen <thijsvanseveren@gmail.com>:

> Hi Paul
> Yesterday i built hydrogen from your fork but i didnt have time to play
> around with it yet
> I'll have a go during my lunch break
>
> Grtz
> Thijs
> Op 25-mrt.-2015 16:52 schreef "Paul Vint" <notifications@github.com>:
>
> I have this partially done, I added an tag to the xml with the filename of
>> the image, and it displays in the properties dialog much like your image
>> above, and is copied to a new drumkit correctly when the kit is saved under
>> a new name.
>>
>> It is *not* in the Instruments->Save dialog yet, nor in the library
>> browser, but I thought I'd let you have a look before I go too much
>> further.
>>
>> You can have a look at the diffs at master...pvint:master
>> <https://github.com/hydrogen-music/hydrogen/compare/master...pvint:master>
>> or I can do a pull request if you'd like.
>>
>> Also note: The TR-808 drumkit is updated and it get's rid of the XML
>> errors.
>>
>> Question: Is the source for feeds/drumkit_list.php available somewhere?
>>
>> --
>> Reply to this email directly or view it on GitHub
>> <https://github.com/hydrogen-music/hydrogen/issues/212#issuecomment-86092880>
>> .
>>
>


-- 
follow me on my Audio & Linux blog <http://audio-and-linux.blogspot.com/> !

**Comment by pvint** [comment]
Good to hear :-)

I have it so that the image is scaled down if need be, but if the image is smaller it shows it as-is. 
As for transparency - I tried it and it works. It will load any image that QT supports, but I only test with png and jpeg.

Getting the images going now for kits from the remote server... the way I have it working now (not committed to git yet) is to use the URL from the server list (ie: hydrogen-music.org/feeds/drumkit_list.php) and looking for the image under that directory, in a "drumkits" directory.

Example: 
Let's say that the Yamaha Vintage kit has an image specified in the xml like:
`<image>yamaha_drums.jpeg</image>`
The image's URL will be http://www.hydrogen-music.org/feeds/drumkits/YamahaVintageKit/yamaha_drums.jpeg

I have no idea what's on that server, but seemed like a reasonable idea to have the kits expanded in a drumkits/ subdirectory. Easy to change if not correct or not desired.

Should have this part tidied up tomorrow and I have my server set up as a drumkit repo that we can test on - I'll let you know when it's ready for testing.

Paul

**Comment by pvint** [comment]
Update: 
Fetching images from remote repository now works in the Import dialog (image is loaded from local disk if the kit is installed). 

TODO: 
- Cache the remote images (Should I just create an "images/" subdirectory under the cache directory for the server?)
- Provide user a way to add image in the properties dialog
- Sort out how images can be stored on server. 

For testing I have a "server" set up at http://dotslash.ca/h2/drumkit_list.xml and the TR808 and the YamahaVintageKit have images set up.

**Comment by pvint** [comment]
- Added caching of remote images. It's simply saving a copy in "data/cache/repositories/images/" directory. Is there maybe a better way to do this? Is it worth worrying about different kit images having the same name?

**Comment by pvint** [comment]
- Instrument Save Library done... 
Barring some testing I think it's done! :-)

**Comment by thijz** [comment]
Hi Paul
So if i understand correctly the image is not part of the drumkit?
Grtz
Thijs
 Op 28-mrt.-2015 22:12 schreef "Paul Vint" <notifications@github.com>:

**Comment by pvint** [comment]
The way I've set it up the image is in with all the other files in the drumkit, and referenced by `<image>` in the xml.

Maybe the confusion is regarding having the images stored on the "drumkit_list.php" server so users get images in Hydrogen when browsing for non-installed kits - I was thinking the easiest way to handle these would be to have images in a directory on the server, but I'm open to ideas.

**Comment by thijz** [comment]
Ok, got it!
That makes sense
I'll try out your test server today

Grtz
Thijs
Op 28-mrt.-2015 23:01 schreef "Paul Vint" <notifications@github.com>:

**Comment by thijz** [comment]
hi Paul
i'm going trough all open tickets so i ended up here
i played around some more and quite like the work you did  :-)

one thing will need fixing and that is the separation between the picture and the text that the drumkit author added
if you select the 'belofilms.com - AC-Guitar-strums' drumkit, you will see that the text on the right is 'cut off' abruptly 
this is (i think?) due to the area that has been reserved for the picture

so i guess we need to split up the picture and text area + add a scrollbar to the text area
see pic below

![image](https://cloud.githubusercontent.com/assets/1384210/9081530/4b509568-3b5d-11e5-95aa-e82f9eb0bd2f.png)

**Comment by thijz** [comment]
just noticed one more thing:
afte i install a drumkit that has a picture (i selected the DeathMetal kit) i expected to see that pic in the properties window of that kit, but i get this : 

![image](https://cloud.githubusercontent.com/assets/1384210/9081658/ad36e5ce-3b5e-11e5-9b6a-bed904849fcc.png)

i can select a pic on my local pc and save the kit, but that's only relevant if you are creating a new kit and want to add a pic to it
for downloaded drumkits that already have a pic, i expect to see the pic that came with the drumkit when i downloaded & installed the kit

does that make sense ?

**Comment by pvint** [comment]
Yeah - the formatting needs some work.... My thinking is to have the picture space expand automagically to the size of the image (up to something like 50% of the space maximum) and add scrollbar to the text.

For the second part - in the properties, I think I've done that but maybe not committed it. I'll have to check... I'm at work now and will be away all weekend, so I'll try to get back to you Monday or Tuesday on this.

Hoping to get this done soon.... it's been a busy summer and I've just moved and started a new job, but I should be able to put some time into this again in the near future.

**Comment by thijz** [comment]
Op 6-aug.-2015 21:52 schreef "Paul Vint" <notifications@github.com>:
>
> Yeah - the formatting needs some work.... My thinking is to have the
picture space expand automagically to the size of the image (up to
something like 50% of the space maximum) and add scrollbar to the text.

Sounds good !

>
> For the second part - in the properties, I think I've done that but maybe
not committed it. I'll have to check... I'm at work now and will be away
all weekend, so I'll try to get back to you Monday or Tuesday on this.
>
Sure, dont worry!
No need to rush

> Hoping to get this done soon.... it's been a busy summer and I've just
moved and started a new job, but I should be able to put some time into
this again in the near future.
>
Good luck with the new house and job!
Sound like you wont be bored any time soon ;-)

**Comment by thijz** [comment]
Hi Paul
did you get a chance to have a look at this yet ?

grtz
Thijs

**Comment by pvint** [comment]
It's getting to the time of year that gives me more time for this, and I've *begun* to revisit it.... the first thing I need to do is get my fork synced with the upstream repo, and there is (as expected) a few collisions that I need to sort out. In a week or so I'll have more time, but in the meantime I'll pick away at it when I can. 

Cheers
Paul

**Comment by pvint** [comment]
- Description text area is now a read-only edittext and is scrollable (not more cut-off text)
- Properties window shows image

Getting closer....

**Comment by pvint** [comment]
Thijs - would you mind giving things a test from my fork? I was hoping to really give it full test this week but been too busy with work and whatnot. It's pretty much there except for figuring out exactly how we want to store images on the server (I'm leaning towards just adding a subdirectory called drumkitimages/ and putting all of the images there using the same basename as the .h2drumkit file (ie: mydrumkit.h2drumkit would reference drumkitimages/mydrumkit.png/jpeg/jpg)
I will really try to give it a good test and finish up this weekend, but time might be a bit tight to get too much done.

Cheers
Paul

**Comment by thijz** [comment]
Hi Paul !
just pulled you fork and building it right now
will report back ASAP

grtz
Thijs

**Comment by thijz** [comment]
some issues i found : 

issues related to the sound library right-click properties window :  

1) imho this should be a 'read only' info screen, since you just want to get info about it.  If you want to edit the drumkit properties you should just create your own copy of it (load the drumkit, Instrument > Save library > edit properties > done)
But since it has always been like that i guess this is not an issue

2) in the properties window you can click the image browse button for the drumkit image > OK
however, if you click browse and then hit cancel in the window that pops up the image field is empty
I would expect it to keep the previous value if you cancel your action

3) i would suggest using a fixed size image area (both in the drumkit properties window and in import library window
That way you are sure that your image will always look good

4) (now i'm getting into the details ;-) the spacing between the information field, image field and image license field are not equal 
![image](https://cloud.githubusercontent.com/assets/1384210/11117665/01647b22-893c-11e5-937c-d3a3401f7ca8.png)
Also the 'image license' text next to the image license field is not completely visible (this could depend on gui settings, but i'm using the default settings)



issues related to saving/editing/creating drumkits
1) if i right-click a drumkit (that has no image) in the library area & open the properties window i can add an image.  If i do this and save, it looks like the image is added to the kit, but if i close hydrogen and reopen it the image is gone

2) save as above (add image to drumkit & save). if you export the drumkit, then delete the drumkit from hydrogen, import the drumkit you just exported the image is not present in the properties window


that's it for now (lunchbreak is over ;-)
i'll test some more in the next days

grtz
Thijs
However, if i then export this drumkit

**Comment by pvint** [comment]
Excellent feedback - I should be able to have a look tomorrow night (Friday).
Thanks Thijs!
Paul

**Comment by pvint** [comment]
Addressed (somewhat) the first few notes from above:

1. "Read only properties" - I'm not worrying about this just now - as you said that's the way it has always been.... can be changed easily enough if it's desired.

2. In Properties dialog - Cancel now really does cancel. (nice catch - thanks)

3. Fixed size for import vs properties window: The way I have it set up the image will get scaled - I think it should stay that way so that the dialogs are more easily converted to scalable.  (I know there's a bug report about a dialog extending beyond the visible screen and dialogs being scalable or scrollable will likely come up). Thoughts?

4. Fixed the spacing on the Properties dialog.

I'll look at the "issues related to saving/editing/creating drumkits" now.

Cheers
Paul

**Comment by pvint** [comment]
On the "issues related to saving/editing/creating drumkits" part - Fixed #1, and that should fix #2 as well (When I merged the upstream changes I lost a line in drumkit.cpp).

It's close  :)

I'll do some more testing tonight/tomorrow.

Good work Thijs, btw - nice testing and feedback.

Cheers!
Paul

**Comment by thijz** [comment]
hi Paul !
referring to your second last reply:
1. and 2. : OK
3. good point, let's leave it like it is
4. spacing looks better, but some of the element are still a couple of pixels off:
![image](https://cloud.githubusercontent.com/assets/1384210/11166039/dd2dee98-8b23-11e5-907a-ddb2d9ac67e9.png)

the browse buton and image location field are not aligned vertically
it would also be nice if the image browse button and the image area could be right-aligned with the other fields ('image license field')

also in the soundlibrary properties windows the 'license' field should be renamed to 'drumkit license' (see remark below)


in the  sound library import dialogue you added a image license field > OK
but maybe it would be more clear if the original 'license' field that refers to the drumkit license was renamed to 'drumkit license' and if it was moved just below the drumkit description field
maybe the 'image license' filed should be placed just below the image
that way it is clear what the license fields refer to

![image](https://cloud.githubusercontent.com/assets/1384210/11166011/853e47ba-8b22-11e5-81c6-08b21ddea688.png)

**Comment by pvint** [comment]
Made changes as noted above (except I didn't move the "drumkit license" up above the image - I think it looks better where it is, although I do agree with the logic of moving it up (or maybe I was just being lazy ;)

I made a bunch of other little refinements, and also repackaged a bunch of kits and threw them up on the server - try it with the import dialog using http://dotslash.ca/h2/list/xml for the server URL.

Note that the info field is completely missing and there's a few oddities like missing names etc, but that's just the XML - I was also working on a script to generate the xml and there's a few issues with it but it's good enough for a quick demo.

/p

**Comment by thijz** [comment]
hi Paul, great work !

import window : drumkit license/image license > looks good and it's indeed clear enough now
properties window : looks great !!  but i noticed that when i imported the denon CRB90 (great pic btw  ;-) that the drumkit was actually imported a second time, or at least was listed twice in my sound library window
also the properties window for both entries did not show a picture
when i tried to delete both entries i got an error when i tried to delete the second entry : "drumkit deletion failed"
after restarting hydrogen i still had 1 entry in my list, and now the properties window did show the picture !?!

i tried the same with the boss dr110 kit (import it a second time from your new drumkit url) and i got the same result : 2 dr110 entries in the list and no picture in the properties window for these kits
as soon as i delete 1 of the 2 entries the picture is displayed in the properties window of the remaining entry
...
strange


as a side note i did have some issues when i clicked some kits (in the import window) from your previous url http://dotslash.ca/h2/list.xml (like the boss dr110) but i guess that is not very relevant ?
![image](https://cloud.githubusercontent.com/assets/1384210/11195299/952b1c8a-8cb1-11e5-945d-03b784b7653f.png)

grtz
Thijs

**Comment by pvint** [comment]
Whups! I should have warned you that kits could get imported twice in some
cases - When I put them on my server I removed the spaces from any of the
filenames and made the .h2drumkit filenames match the name of the directory
in the tarball. (I think we want to update the ones on the SF servr like
that, correct?). I guess strange things might happen if you have installed
two kits with a different .h2drumkit name but have the same directory
name.... perhaps if the kits are updated so the directories and .h2drumkit
names match we will have to make them different than the old names to avoid
collisions like that (or maybe we just need to make H2 handle the spaces
and directory/filename differences better?)

Anyway, long story short: The errors you got are just because of my
incomplete xml on the test server combined with drumkits with same
directory but (slightly) different names.

PS: I'm not sure what the licensing of the image I used for the CRB90 is,
but I had a good feeling that it was OK to use. ;)

I don't have time tonight, but tomorrow night I can give it a good test and
check everything and then maybe it's finally time for a pull request!

Cheers, and thanks for the testing!
Paul

**Comment by trebmuh** [comment]
> I think we want to update the ones on the SF servr like that, correct?
I think that the way to go as it's discussed here : https://github.com/hydrogen-music/hydrogen/issues/298
By the way, well done for this work :)

**Comment by pvint** [comment]
Added pull request #334 on this. 
There's likely a couple kins to get out of it, but there it is, and many many thanks to a few, but especially @thijz for the testing and the great feedback.

**Comment by mauser** [comment]
This is now part of the main branch, so i'm going to close the issue..
